third_party/move/move-compiler-v2/src/env_pipeline/closure_checker.rs

```diff
@@ -17,9 +17,49 @@ use move_core_types::ability::Ability;
 use move_model::{
     ast::{ExpData, Operation},
     model::GlobalEnv,
+    ty::Type,
     well_known,
 };
 
+/// Check if the type contains an option type
+fn contains_option_type(env: &GlobalEnv, ty: &Type) -> bool {
+    match ty {
+        Type::Vector(ty) => contains_option_type(env, ty),
+        Type::Struct(mid, sid, tys) => {
+            let struct_env = env.get_module(*mid).into_struct(*sid);
+            if struct_env.get_full_name_with_address() == "0x1::option::Option" {
+                return true;
+            }
+            if struct_env.has_variants() {
+                for variant in struct_env.get_variants() {
+                    for field in struct_env.get_fields_of_variant(variant) {
+                        if contains_option_type(env, &field.get_type()) {
+                            return true;
+                        }
+                    }
+                }
+            } else {
+                for field in struct_env.get_fields() {
+                    if contains_option_type(env, &field.get_type()) {
+                        return true;
+                    }
+                }
+            }
+            tys.iter().any(|ty| contains_option_type(env, ty))
```

Comment:
Are these type arguments? If they are phantom we can/should allow option? Corner case though

Reply:
I think this part is already covered by the branch `Type::TypeParameter(..) => false`?

Reply:
yeah, forgot about that. Updated the PR